zipkin-ui/README.md

```diff
@@ -0,0 +1,34 @@
+# zipkin-ui
+
+Zipkin-UI is a single-page application that reads configuration from `/config.json`.
+
+When looking at a trace, the browser is sent to the path `/traces/{id}`. For the single-page
+app to serve that route, the server needs to forward the request to `index.html`. The same
+forwarding applies to `/dependencies` and any other routes the UI controls. 
+
+Under the scenes the JavaScript code looks at `window.location` to figure out what the
+UI should do. This is handled by a route api defined in the crossroads library.
+
+The full list of
+such routes is defined in [`resources/zipkin-ui/index-routes.json`](src/main/resources/zipkin-ui/index-routes.json).
+Routes ending with `/` should be used in prefix matching (`/foo/` should match `/foo/`, `/foo/bar`, and `/foo/bar/baz`).
+Routes not ending with `/` should be used in exact matching (`/foo` should match `/foo` and nothing else).
+
+Based on all these, there are several ways for serving zipkin-ui:
+
+ * Programatically read `index-routes.json` and set up explicit routes based on it. Example: [zipkin-web](https://github.com/openzipkin/zipkin/blob/5dec252e4c562b21bac5ac2f9d0b437d90988f79/zipkin-web/src/main/scala/com/twitter/zipkin/web/Main.scala#L92-L95).
```

Comment:
Watch out, wall-of-text incoming ;)

Serving index.html is possible with any server; that's exactly what's documented in the second option. The question is more whether we want to do that. In the end, all users of zipkin-ui are free to do with the package whatever they want, what we need to decide is the use-cases we support / document.

To me "serve index.html always" feels icky; if we know something's not there, I have a slight preference to let the server already tell me so. Not doing so breaks strong conventions. I've already seen several times that it makes debugging harder (I look at the network tab of my browser, see 200s, but nothing works; I guess that one of the JS files is wrong, click through, and _then_ see that the contents of `app.min.js` are the contents of `index.html`, _and then_ from that infer that this is in fact a 404, and to verify that, I'll need to open the URL in a new browser tab)

Those are my arguments for "support this use-case".

Arguments I see so far against "support this use-case": I see that maintaining `index-routes.json` is extra work. Based on what I've seen so far, I expect that to mean adding an extra line every few months at most. I understand that defaulting to `index.html` is the way many modern single-page apps are build. Dropping this use-case also does away with the extra complexity in users that'd use `index-routes.json`, for the price of not allowing them to implement that extra complexity, even if they want to. (I'd want to, but today I'm playing maintainer of Zipkin as opposed to user, so that's mostly off-topic)

Did I miss any arguments on either side? If no: based on these, I have a weak preference to provide `index-routes.json`. I understand your evaluation is more firmly on the other side. I'd love to get more opinions.

PS. I'm totally OK getting rid of the relevant code if we decide to not support `index-routes.json`, that "wasted work" should not be a factor here.